endpoints/openrtb2/auction.go

```diff
@@ -1582,44 +1583,76 @@ func getJsonSyntaxError(testJSON []byte) (bool, string) {
 // because stored imps and stored requests can contain stored auction responses
 // so the stored requests/imps have to be merged into the incoming request prior to processing stored auction responses.
-func (deps *endpointDeps) processStoredAuctionResponses(ctx context.Context, requestJson []byte) (map[string]json.RawMessage, []error) {
+func (deps *endpointDeps) processStoredAuctionResponses(ctx context.Context, requestJson []byte) (map[string]json.RawMessage, map[string]map[string]json.RawMessage, []error) {
 	impInfo, errs := parseImpInfo(requestJson)
 	if len(errs) > 0 {
-		return nil, errs
+		return nil, nil, errs
 	}
 
 	storedAuctionResponseIds := make([]string, 0, 0) //all stored responses ids from all imps
 	//because of bulk fetch responses we need to map imp id to stored resp body
 	impIdToRespId := make(map[string]string)              //imp id to stored resp id
 	impIdToStoredResp := make(map[string]json.RawMessage) //imp id to stored resp body
+
+	//stored bid responses: imp id to bidder to stored response id
+	impBidderToStoredBidResponseId := make(map[string]map[string]string)
+	//stored bid responses: imp id to bidder to stored response body
+	impBidderToStoredBidResponse := make(map[string]map[string]json.RawMessage)
+
 	for index, impData := range impInfo {
 
+		impId, err := jsonparser.GetString(impData.Imp, "id")
+		if err != nil {
+			return nil, nil, []error{fmt.Errorf("request.imp[%d] missing required field: \"id\"", index)}
+		}
+
 		if impData.ImpExtPrebid.StoredAuctionResponse != nil {
 			if len(impData.ImpExtPrebid.StoredAuctionResponse.ID) == 0 {
-				return nil, []error{fmt.Errorf("request.imp[%d] has ext.prebid.storedauctionresponse specified, but \"id\" field is missing ", index)}
+				return nil, nil, []error{fmt.Errorf("request.imp[%d] has ext.prebid.storedauctionresponse specified, but \"id\" field is missing ", index)}
 			}
 			storedAuctionResponseIds = append(storedAuctionResponseIds, impData.ImpExtPrebid.StoredAuctionResponse.ID)
 
-			impId, err := jsonparser.GetString(impData.Imp, "id")
-			if err != nil {
-				return nil, []error{err}
-			}
-
 			impIdToRespId[impId] = impData.ImpExtPrebid.StoredAuctionResponse.ID
 
 		}
+		if len(impData.ImpExtPrebid.StoredBidResponse) > 0 {
+
+			bidderStoredRespId := make(map[string]string)
+			for _, bidderResp := range impData.ImpExtPrebid.StoredBidResponse {
+				if len(bidderResp.ID) == 0 || len(bidderResp.Bidder) == 0 {
+					return nil, nil, []error{fmt.Errorf("request.imp[%d] has ext.prebid.storedbidresponse specified, but \"id\" or/and \"bidder\" fields are missing ", index)}
+				}
+				//check if bidder is valid/exists
+				if _, isValid := deps.bidderMap[bidderResp.Bidder]; !isValid {
+					return nil, nil, []error{fmt.Errorf("request.imp[impId: %s].ext contains unknown bidder: %s. Did you forget an alias in request.ext.prebid.aliases?", impId, bidderResp.Bidder)}
+				}
+				//assuming bidder is unique per one bid stored response
```

Comment:
Is this comment necessary? Just checking :)

Reply:
I'm 90% sure this should be unique, but I need to check with Bret. I'll remove the comment when I get the answer.

Reply:
Per discussion with Bret we don't need to support this. I'll remove a comment on my next commit.